File: CFGCalculator/src/Parser.java
import java.util.Scanner;

public class Parser {

    int cursor = 0;

    String input;

    String buff;


    public Parser(String input){
        this.input = input+"$";
        cursor = 0;
        buff = "    ";
    }

    public double calculate(){
        return parseExpression("");
    }

    public double parseExpression(String buffer){
        System.out.println(buffer+"Expression");
        buffer += buff;
        double head = parseTerm(buffer);
        double tail = parseExpressionTail(buffer);
        return head + tail;
    }

    public double parseExpressionTail(String buffer){
        System.out.println(buffer+"Exp_Tail");
        buffer += buff;
        char c = peek();
        if(c == '+' || c == '-'){
            int sign = consumeAddOp(buffer);
            double head = parseTerm(buffer);
            double tail = parseExpressionTail(buffer);
            return sign * head + tail;
        }else{
            return 0;
        }

    }


    public double parseTerm(String buffer){
        System.out.println(buffer+"Term");
        buffer += buff;
        double head = parseFactor(buffer);
        double tail = parseTermTail(buffer);
        return head * tail;
    }

    public double parseTermTail(String buffer){
        System.out.println(buffer+"Term_Tail");
        buffer += buff;
        char c = peek();
        if(c == '*' || c == '/') {
            char sign = consumeMultiOp(buffer);
            double head = parseFactor(buffer);
            if (sign == '*') {

            } else if (sign == '/') {
                head = 1 / head;
            }

            double tail = parseTermTail(buffer);

            return head * tail;
        }else{
            return 1;
        }
    }

    public double parseFactor(String buffer){
        System.out.println(buffer+"Factor");
        buffer += buff;
        if(cursor >= input.length()){
            return 1;
        }

        char c = peek();

        if(c=='(') {
            cursor++;
            double val = parseExpression(buffer);
            cursor++;
            return val;

        }else if(isDigit(c)) {

            double head = parseNumber(buffer);

            return head;

        }else{
            return 1;
        }
    }

//    public double parseFactorTail(){
//        char cur = peek();
//        cursor++;
//
//
//        switch(cur){
//            case '*':
//        }
//        return 0;
//    }

    public int consumeAddOp(String buffer){
        System.out.println(buffer+"AddOP");

        char c = peek();
        cursor++;
        if(c=='+'){
            return 1;
        }else {
            return -1;
        }
    }

    public char consumeMultiOp(String buffer){
        System.out.println(buffer+"MulOP");

        char c = peek();
        cursor++;
        return c;
    }



    public double parseNumber(String buffer){
        System.out.println(buffer+"Number");

        String res = "";
        while(cursor < input.length()){
            char c = peek();
            if(isDigit(c)){
                res += c;
                cursor++;
            }else{
                break;
            }
        }
        return Double.parseDouble(res);
    }


    public char peek(){
        return input.charAt(cursor);
    }

    public static boolean isDigit(char c){
        int i = c;
        return (i >= 48) && (i <= 57);
    }
    public static void main(String[] args){
        Scanner scan = new Scanner(System.in);
        System.out.println("Please enter your expression:");
        String str = scan.nextLine();
        Parser parser = new Parser(str);
        parser.buff = "          ";
        try{
            System.out.println("\nLegal Expression, start to calculate:\n");
            double result = parser.calculate();

            System.out.println("\nThe result is: " + result);
        }catch(Exception e){
            System.out.println("\nIllegal Expression, exit.");
        }
        scan.close();
    }
}

File: CFGCalculator/parser.py

class Parser:

	def __init__(self, input, cursor):
		self.input = input
		self.cursor = cursor

	def setInput(self, input):
		self.input = input + "e"

	def parseExpression(self):
		term = self.parseTerm()
		tail = self.parseExpressionTail()
		return term + tail
	
	def parseExpressionTail(self):
		if(not isAddOp(self.peek())):
			return 0

		op = self.parseAddOp()
		term = self.parseTerm()
		if(op == "-"):
			term = 0 - term
		tail = self.parseExpressionTail()

		return term + tail
	
	def parseTerm(self):
		head = self.parseFactor()
		tail = self.parseTermTail()
		return head * tail
	
	def parseTermTail(self):
		if(not isMultiOp(self.peek())):
			return 1

		op = self.parseMultiOp()
		factor = self.parseFactor()
		if(op == "/"):
			factor = 1.0 / factor
		tail = self.parseTermTail()

		return factor * tail

	def parseFactor(self):
		value = None
		if(self.peek() == "("):
			self.match("(")
			value = self.parseExpression()
			self.match(")")
		else:
			value = self.parseNumber()
		return value

	def parseNumber(self):
		res = ""
		while self.cursor < len(self.input):
			if(isDigit(self.input[self.cursor])):
				res += self.freeMatch()
			else:
				break
		return int(res)
	
	def parseAddOp(self):
		token = self.peek()
		if(token == "+" or token == "-"):
			return self.freeMatch()
		else:
			return None

	def parseMultiOp(self):
		token = self.peek()
		if(token == "*" or token == "/"):
			return self.freeMatch()
		else: 
			return None

	# match current character
	def match(self, char):
		if self.input[self.cursor] == char:
			self.cursor += 1
			return char
		else:
			return None
	
	# free match current character
	def freeMatch(self):
		res = self.peek()
		self.cursor += 1
		return res

	def peek(self):
		return self.input[self.cursor]

	def consumeSpace(self):
		while self.cursor < len(self.input):
			if self.peek() == " ":
				self.freeMatch()
			else:
				break


def new_Parser(input):
	return Parser(input, 0)

def isDigit(input):
	return ord(input) >= 48 and ord(input) <= 57

def isMultiOp(char):
	return char == "*" or char == "/"

def isAddOp(char):
	return char == "+" or char == "-"
	

exp = "5-(1+2*3)-(4*5/2)+100"
parser = new_Parser(exp)


print(parser.parseExpression())


